File: iOS/SDK/smartsip-sdk/smartsip-sdk/Internal/linphone/PhoneCore.swift
//
//  LinPhoneCore.swift
//  smartsip-sdk
//
//  Created by Franz Iacob on 09/01/2026.
//

internal import linphonesw
import os
import Foundation

internal final class PhoneCore {
    
    private var mCore: Core!
    private var phoneCoreDelegate: CallDelegate?
    private var callInfo: CallInfo!

    internal required init()
    {
        do
        {
            mCore = try Factory.Instance.createCore(configPath: "", factoryConfigPath: "", systemContext: nil)
            //disable video
            mCore.videoCaptureEnabled = false
            mCore.videoDisplayEnabled = false
            mCore.videoActivationPolicy!.automaticallyAccept = false
            //force encryption to Secure RTP
            try mCore.setMediaencryption(newValue: .SRTP)
            
            //feed the state to the core listening for changes, if any
            let mCoreDelegate = CoreDelegateStub( onCallStateChanged: { (core: Core, call: Call, state: Call.State, message: String) in
                
                if( state == .Error )
                {
                    self.phoneCoreDelegate?.callDidFail(withError: message);
                }
                else
                {
                    self.phoneCoreDelegate?.callDidChangeState(CallState.from(linphoneState: state) ?? CallState.disconnected)
                }
                
            }, onAccountRegistrationStateChanged: { (core: Core, account: Account, state: RegistrationState, message: String) in
                
                if( state == .Failed )
                {
                    self.phoneCoreDelegate?.callDidFail(withError: message)
                }
                else
                {
                    if ( state == .Ok )
                    {
                        Logger.sdk.info("✅ Registered Successfully. Auto-starting the call...")
                        // This is where we trigger the call immediately after login
                        self.makeOutgoingCall()
                    }
                    self.phoneCoreDelegate?.callDidChangeState(CallState.from(linphoneState: state) ?? CallState.loggedOut)
                }
                
            })
            mCore.addDelegate(delegate: mCoreDelegate)
        }
        catch {
            Logger.sdk.error("❌ initializeDependencies linphonesw error: \(error.localizedDescription)")
        }
    }
    
    internal func setDelegate(_ delegate: CallDelegate) {
        self.phoneCoreDelegate = delegate
    }
    
    internal func makeCall(callInfo: CallInfo!) async
    {
        self.callInfo = callInfo;
        Logger.sdk.info("🚀 Initiating SIP Call...")
        Logger.sdk.info("SIP Server: \(callInfo.domain):\(callInfo.port)")
        Logger.sdk.info("SIP User: \(callInfo.username)")
        
        //login into the SIP server
        let callerAddress : String = "sip:\(callInfo.sessionId)@\(callInfo.domain)"
        
        do
        {
            // Configure the SIP Authentication credentials using the session results
            // We use the username, password, and domain (realm) provided by the middleware
            let authInfo = try Factory.Instance.createAuthInfo(
                username: callInfo.username,
                userid: "",
                passwd: callInfo.password,
                ha1: "",
                realm: "",
                domain: callInfo.domain
            )

            // Initialize account parameters to define the user identity and registration behavior
            let accountParams = try mCore.createAccountParams()

            // Create and set the 'From' identity address (e.g., sip:username@domain)
            let identity = try Factory.Instance.createAddress(addr: callerAddress)
            try accountParams.setIdentityaddress(newValue: identity)

            // Configure the Remote Proxy/Server address for the SIP connection
            // We use SIPS (Secure SIP) over TLS via port 443 as required by the infrastructure
            let address = try Factory.Instance.createAddress(addr: "sips:\(callInfo.domain)")
            try address.setTransport(newValue: TransportType.Tls)
            try address.setPort(newValue: callInfo.port)
            try accountParams.setServeraddress(newValue: address)

            // Enable registration so the client sends a REGISTER request to the proxy
            accountParams.registerEnabled = true

            // Instantiate the account with the defined parameters
            let account = try mCore.createAccount(params: accountParams)

            // Add authentication info to the Core cache and register the account
            // Setting the account as 'default' ensures outgoing calls use this identity
            mCore.addAuthInfo(info: authInfo)
            try mCore.addAccount(account: account)
            mCore.defaultAccount = account
            
            // Starts the SIP stack main loop and begins processing background network tasks.
            // This initiates the actual registration process and prepares the SDK to handle
            // signaling, media streams, and events.
            try mCore.start()
        }
        catch
        {
            Logger.sdk.error("❌ login into SIP server error: \(error.localizedDescription)")
        }
        Logger.sdk.info("🔐 Login in progress with SessionID: \(callInfo.sessionId)")
    }
    
    
    internal func makeOutgoingCall() {
        do {
            Logger.sdk.info("📞 Calling Queue: [SessionID: \(self.callInfo.sessionId)]")
            
            // Construct the SIP URI for the destination
            let remoteAddress = try Factory.Instance.createAddress(addr: "sip:queue@\(self.callInfo.domain)")
            
            // Configure call parameters
            // Passing 'nil' to createCallParams signifies a new outgoing call session
            let params = try mCore.createCallParams(call: nil)
            
            // Attach custom headers for server-side tracking (e.g., SmartSip routing)
            params.addCustomHeader(headerName: "X-SmartSip-session", headerValue: self.callInfo.sessionId)
            
            // Initiate the INVITE request
            _ = mCore.inviteAddressWithParams(addr: remoteAddress, params: params)
            
            // Note: Monitor 'onCallStateChanged' in the CoreListener to track progress
        } catch {
            Logger.sdk.error("❌ Outgoing call failed: \(error.localizedDescription)")
        }
    }
    

    internal func terminateCallAndLogout() {
        do {
            // Hang up the call if it exists
            if mCore.callsNb > 0 {
                let call = mCore.currentCall ?? mCore.calls.first
                try call?.terminate()
                Logger.sdk.info("📞 Call terminated by user.")
            }
            
            // Unregister and clear to ensure no "ghost" registrations remain
            // This sends a REGISTER with expires=0 to the server
            mCore.clearAccounts()
            mCore.clearAllAuthInfo()
            
            Logger.sdk.info("🚪 Session closed and logged out.")
            
            // Update UI to initial state
            self.phoneCoreDelegate?.callDidChangeState(.loggedOut)
            
        } catch {
            Logger.sdk.error("❌ Error during teardown: \(error.localizedDescription)")
        }
    }
    
    /// Configures the verbosity of the underlying SIP stack logs.
    /// - Parameter enabled: If true, the SDK will output detailed debug information.
    ///   If false, only critical errors will be logged.
    internal func setDebugMode(enabled: Bool) {
        // Set the global SIP log level.
        // .Debug provides full trace; .Error restricts output to failures only.
        LoggingService.Instance.logLevel = enabled ? .Debug : .Error
    }
    
    
}

File: Android/smartsip-sdk/src/main/java/cc/smartconnect/smartsip_sdk/private/PhoneCore.kt
package cc.smartconnect.smartsip_sdk.private

import android.content.Context
import cc.smartconnect.smartsip_sdk.CallListener
import cc.smartconnect.smartsip_sdk.CallState
import cc.smartconnect.smartsip_sdk.DTMFButton
import org.linphone.core.*

/**
 * PhoneCore.kt
 * smartsip-sdk
 *
 * Created by Franz Iacob on 21/01/2026.
 * * The central engine responsible for managing the Linphone Core lifecycle,
 * SIP registration, and call signaling.
 */
internal class PhoneCore(private val context: Context) {

    private lateinit var mCore: Core
    private var phoneCoreDelegate: CallListener? = null
    private var callInfo: CallInfo? = null

    init {
        try {
            // Initialize Linphone Core with the Android Application Context
            mCore = Factory.instance().createCore(null, null, context)

            // Video Configuration: Explicitly disabled for voice-only optimization
            mCore.isVideoCaptureEnabled = false
            mCore.isVideoDisplayEnabled = false

            mCore.config.apply {
                setBool("video", "enabled", false)
                setBool("video", "capture", false)
                setBool("video", "display", false)
            }

            // Security: Force Media Encryption to Secure RTP (SRTP)
            mCore.setMediaEncryption(MediaEncryption.SRTP)
            //disable the verification CN of the TLS
            mCore.verifyServerCn(false);

            // Setup Core Listener to handle signaling events
            val coreListener = object : CoreListenerStub() {
                override fun onCallStateChanged(core: Core, call: Call, state: Call.State?, message: String) {
                    if (state == Call.State.Error) {
                        phoneCoreDelegate?.callDidFail(message)
                    } else {
                        val mappedState = CallState.from(state) ?: CallState.DISCONNECTED
                        phoneCoreDelegate?.callDidChangeState(mappedState)
                    }
                }

                override fun onAccountRegistrationStateChanged(
                    core: Core,
                    account: Account,
                    state: RegistrationState?,
                    message: String
                ) {
                    when (state) {
                        RegistrationState.Failed -> {
                            phoneCoreDelegate?.callDidFail(message)
                        }
                        RegistrationState.Ok -> {
                            Logger.sdk.info("✅ Registered Successfully. Auto-starting the call...")
                            // Trigger outgoing call immediately upon successful registration
                            makeOutgoingCall()
                        }
                        else -> {
                            val mappedState = CallState.from(state) ?: CallState.LOGGED_OUT
                            phoneCoreDelegate?.callDidChangeState(mappedState)
                        }
                    }
                }
            }

            mCore.addListener(coreListener)

        } catch (e: Exception) {
            Logger.sdk.error("❌ initializeDependencies linphone error: ${e.localizedMessage}")
        }
    }

    // --- Internal API ---

    internal fun setDelegate(delegate: CallListener) {
        this.phoneCoreDelegate = delegate
    }

    internal suspend fun makeCall(callInfo: CallInfo) {
        this.callInfo = callInfo

        Logger.sdk.info("🚀 Initiating SIP Call...")
        Logger.sdk.info("SIP Server: ${callInfo.domain}:${callInfo.port}")
        Logger.sdk.info("SIP User: ${callInfo.username}")

        val callerAddress = "sip:${callInfo.sessionId}@${callInfo.domain}"

        try {
            // Authentication Setup
            val authInfo = Factory.instance().createAuthInfo(
                callInfo.username,
                "",
                callInfo.password,
                "",
                "",
                callInfo.domain
            )

            // Account Configuration
            val accountParams = mCore.createAccountParams()
            val identity = Factory.instance().createAddress(callerAddress)
            accountParams.identityAddress = identity

            val address = Factory.instance().createAddress("sips:${callInfo.domain}")
            address?.transport = TransportType.Tls
            address?.port = callInfo.port
            accountParams.serverAddress = address
            accountParams.isRegisterEnabled = true

            val account = mCore.createAccount(accountParams)

            mCore.addAuthInfo(authInfo)
            mCore.addAccount(account)
            mCore.defaultAccount = account

            // Start the Core engine
            mCore.start()

        } catch (e: Exception) {
            Logger.sdk.error("❌ login into SIP server error: ${e.localizedMessage}")
        }
        Logger.sdk.info("🔐 Login in progress with SessionID: ${callInfo.sessionId}")
    }

    internal fun makeOutgoingCall() {
        val info = callInfo ?: return
        try {
            Logger.sdk.info("📞 Calling Queue: [SessionID: ${info.sessionId}]")

            val destination = info.destination ?: "queue"
            val remoteAddress = Factory.instance().createAddress("sip:$destination@${info.domain}")

            val params: CallParams? = mCore.createCallParams(null)

            // Inject custom session tracking header
            params?.addCustomHeader("X-SmartSip-session", info.sessionId)

            if (params != null && remoteAddress != null) {
                mCore.inviteAddressWithParams(remoteAddress, params)
            } else {
                Logger.sdk.error("❌ Call initialization failed: Invalid params or address")
            }

        } catch (e: Exception) {
            Logger.sdk.error("❌ Outgoing call failed: ${e.localizedMessage}")
        }
    }

    internal fun terminateCallAndLogout() {
        try {
            val currentCall = mCore.currentCall

            if (currentCall != null) {
                Logger.sdk.info("📞 SIP: Terminating active call session (State: ${currentCall.state})")
                currentCall.terminate()
            } else {
                Logger.sdk.info("📞 SIP: No active call to terminate, stopping all calls.")
                mCore.terminateAllCalls()
            }

            mCore.clearAccounts()
            mCore.clearAllAuthInfo()

            Logger.sdk.info("🚪 Session closed and logged out.")
            this.phoneCoreDelegate?.callDidChangeState(CallState.LOGGED_OUT)

        } catch (e: Exception) {
            Logger.sdk.error("❌ Error during teardown: ${e.localizedMessage}")
        }
    }

    // --- Media Management ---

    fun setSpeakerOn(isSpeakerOn: Boolean) {
        val devices = mCore.audioDevices
        val targetType = if (isSpeakerOn) AudioDevice.Type.Speaker else AudioDevice.Type.Earpiece
        val targetDevice = devices.find { it.type == targetType }

        if (targetDevice != null) {
            // Apply routing to both the engine and the active session
            mCore.outputAudioDevice = targetDevice
            mCore.currentCall?.outputAudioDevice = targetDevice
            Logger.sdk.info("🔊 Audio Routing: Switched to ${targetDevice.deviceName}")
        } else {
            Logger.sdk.error("❌ Audio Routing: Could not find device of type $targetType")
        }
    }

    fun setMicrophoneMuted(muted: Boolean) {
        val currentCall = mCore.currentCall
        if (currentCall != null) {
            currentCall.microphoneMuted = muted
            Logger.sdk.info("🎙️ Microphone ${if (muted) "Muted" else "Unmuted"} for active call.")
        } else {
            // Fallback: update core preference if no call is present
            mCore.isMicEnabled = !muted
        }
    }

    internal fun sendDTMF(tone: DTMFButton) {
        val currentCall = mCore.currentCall ?: run {
            Logger.sdk.error("DTMF: No active call found.")
            return
        }

        val toneChar = tone.rawValue.firstOrNull() ?: return

        try {
            currentCall.sendDtmf(toneChar)
            Logger.sdk.debug("DTMF: Tone '$toneChar' successfully sent.")
        } catch (e: Exception) {
            Logger.sdk.error("DTMF: Failed to send '$toneChar': ${e.localizedMessage}")
        }
    }

    internal fun setNetworkReachable(reachable: Boolean) {
        mCore.isNetworkReachable = reachable
    }

    /**
     * Toggles the SIP Hold state by sending a Re-INVITE to the proxy.
     * @param hold If true, sends a=sendonly; if false, sends a=sendrecv.
     */

    internal class PhoneCore(private val context: Context) {
        private lateinit var mCore: Core
        // ... existing init logic ...

        /**
         * Toggles the SIP Hold state and HARD silences the hardware.
         * @param hold If true, sends SIP Re-INVITE and kills the local audio path.
         */
        fun setCallOnHold(hold: Boolean) {
            val currentCall = mCore.currentCall ?: return

            if (hold) {
                Logger.sdk.info("📞 SIP: Executing Hard Hold (Signaling + Device Detach)")

                // 1. Signaling: Sends Re-INVITE with a=sendonly
                currentCall.pause()

                // 2. Local Media: Disable the microphone capture
                mCore.isMicEnabled = false

                // 3. Hardware Isolation: Disconnect the audio device from the engine.
                // This is the only way to guarantee the speaker stops on some Android builds.
                mCore.outputAudioDevice = null
            } else {
                Logger.sdk.info("📞 SIP: Resuming (Signaling + Device Attach)")

                // 1. Re-attach Hardware: Find the default speaker/earpiece
                val defaultDevice = mCore.audioDevices.firstOrNull {
                    it.type == AudioDevice.Type.Earpiece || it.type == AudioDevice.Type.Speaker
                }
                mCore.outputAudioDevice = defaultDevice

                // 2. Enable Media Capture
                mCore.isMicEnabled = true

                // 3. Signaling: Resume the session
                currentCall.resume()
            }
        }
    }

    internal fun setDebugMode(enabled: Boolean) {
        Factory.instance().setDebugMode(enabled, "SmartSipSDK")
    }
}